pandas/core/indexing.py

```diff
@@ -2016,6 +2016,10 @@ def _setitem_single_column(self, loc: int, value, plane_indexer):
                 and (
                     np.shares_memory(new_values, orig_values)
                     or new_values.shape != orig_values.shape
+                    or (
+                        not can_hold_element(orig_values, np.nan)
+                        and isna(new_values).any()
```

Comment:
how do we get here with warn=True?

Reply:
``iloc._setitem_with_indexer(indexer, value, self.name)``

Reply:
OK, so i think what's happening here is that we are checking can_hold_element too soon, bc reindexing occurs within `self.obj._iset_item(loc, value)`.  I think it would be better to do the reindex before the can_hold_element check

Reply:
Pushed a suggestion. Not sure what is better performance wise. Checking initially and the rechecking? Is ``_get_column_array`` expensive or cheap? If it is cheap, it might be better only checking can_hold_element for new values?

This breaks one test, not sure if you can set an all NaT Series into an underlying Series with timezone ``UTC``?

Reply:
_get_column_array is pretty cheap.  its the isna(...).any() that i think may be expensive

Reply:
> This breaks one test, not sure if you can set an all NaT Series into an underlying Series with timezone UTC?

depends on the dtype of the all-NaT Series

Reply:
original values are ``datetime64[ns, UTC]`` and all NaT Series is ``datetime64[ns]``

Reply:
that should definitely not be inplace

Reply:
Great, than we fixed the test instead of breaking it :)

Anything against checking only new_values then? See latest commit now.

Reply:
i think that makes sense.  caffeine is still kicking in though